src/asset/mod.rs

```diff
@@ -0,0 +1,143 @@
+pub mod audio;
+pub mod image;
+
+pub use audio::Audio;
+pub use image::Image;
+
+use std::ops::Index;
+use std::path::PathBuf;
+use std::time::SystemTime;
+
+/// A marker type for the unit pixels.
+pub type Pixels = usize;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ImageAssetID(usize);
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct AudioAssetID(usize);
+
+/// If the type of asset type is unknown or doesn't matter.
+pub enum AssetID {
+    Image(ImageAssetID),
+    Audio(AudioAssetID),
+}
+
+pub struct AssetSystem {
+    images: Vec<Image>,
+    audio: Vec<Audio>,
+}
+
+impl AssetSystem {
+    pub fn new() -> Self {
+        Self {
+            images: Vec::new(),
+            audio: Vec::new(),
+        }
+    }
+
+    /// Load a new image from disk.
+    pub fn load_image(&mut self, file: PathBuf) -> ImageAssetID {
+        let id = self.images.len();
+        self.images.push(Image::new(file));
+        ImageAssetID(id)
+    }
+
+    /// Load a new sound from disk.
+    pub fn load_audio(&mut self, file: PathBuf) -> AudioAssetID {
+        let id = self.audio.len();
+        self.audio.push(Audio::new(file));
+        AudioAssetID(id)
+    }
+}
+
+impl Index<ImageAssetID> for AssetSystem {
+    type Output = Image;
+
+    fn index(&self, id: ImageAssetID) -> &Self::Output {
+        self.images.get(id.0).expect(&format!("Invalid image asset {}", id.0))
+    }
+}
+
+impl Index<AudioAssetID> for AssetSystem {
+    type Output = Audio;
+
+    fn index(&self, id: AudioAssetID) -> &Self::Output {
+        self.audio.get(id.0).expect(&format!("Invalid audio asset {}", id.0))
+    }
+}
+
+/// Gets an asset from the asset system.
+///
+/// While this might look a bit bloated, it accomplishes two things.
+///
+/// 1) You can only get (e.g.) an [Image] from an [ImageAssetID].
+/// 2) You can always `get` the asset, regardless of its type. Otherwise you'd have
+/// `assets.image(image)`
+pub trait TryGetAsset<O, I> {
+    /// A reference to a loaded asset.
+    fn try_get(&self, id: I) -> Option<&O>;
+}
+
+impl TryGetAsset<Image, ImageAssetID> for AssetSystem {
+    fn try_get(&self, id: ImageAssetID) -> Option<&Image> {
+        self.images.get(id.0)
+    }
+}
+
+impl TryGetAsset<Audio, AudioAssetID> for AssetSystem {
+    fn try_get(&self, id: AudioAssetID) -> Option<&Audio> {
+        self.audio.get(id.0)
+    }
+}
```

Comment:
I'm split about including this. Due to reasons far above my type system knowledge, we can't have the index operator above return a `Option<&T>`, so this is a way of getting an asset and not panicking. But that only makes sense with dynamic loading of assets. So, keep or no? I guess it doesn't hurt anyone and we can still index to get assets unconditionally.

Reply:
Decided to remove.